qiskit/pulse/parameter_manager.py

```diff
@@ -0,0 +1,422 @@
+        raise NotImplementedError
+
+
+class ParameterSetter(NodeVisitor):
+    """Node visitor for parameter binding.
+
+    This visitor is initialized with a dictionary of parameters to be assigned,
+    and assign values to operands of nodes found.
+    """
+    def __init__(self, param_map: Dict[ParameterExpression, ParameterValueType]):
+        self._param_map = param_map
+
+    # Top layer: Assign parameters to programs
+
+    def visit_ScheduleBlock(self, node: ScheduleBlock):
+        """Visit ``ScheduleBlock``. Recursively visit context blocks and overwrite.
+
+        .. note:: ``ScheduleBlock`` can have parameters in blocks and its alignment.
+        """
+        # accessing to protected member
+        node._blocks = [self.visit(block) for block in node.instructions]
+        node._alignment_context = self.visit(node.alignment_context)
+
+        self._update_parameter_manager(node)
+        return node
+
+    def visit_Schedule(self, node: Schedule):
+        """Visit ``Schedule``. Recursively visit schedule children and overwrite."""
+        # accessing to private member
+        node._Schedule__children = [(t0, self.visit(sched)) for t0, sched in node.instructions]
+
+        # update timeslots
+        for chan in copy(node.channels):
+            if isinstance(chan.index, ParameterExpression):
+                chan_timeslots = node._timeslots.pop(chan)
+                new_channel = self.visit_Channel(chan)
+
+                # Merge with existing channel
+                if new_channel in node.timeslots:
+                    sched = Schedule()
+                    sched._timeslots = {new_channel: chan_timeslots}
+                    node._add_timeslots(0, sched)
+                # Or add back under the new name
+                else:
+                    node._timeslots[new_channel] = chan_timeslots
+
+        self._update_parameter_manager(node)
+        return node
+
+    def visit_AlignmentKind(self, node: AlignmentKind):
+        """Assign parameters to block's ``AlignmentKind`` specification."""
+        new_parameters = tuple(self.visit(param) for param in node._context_params)
+        node._context_params = new_parameters
+
+        return node
+
+    # Mid layer: Assign parameters to instructions
+
+    def visit_Call(self, node: instructions.Call):
+        """Assign parameters to ``Call`` instruction.
+
+        .. note:: ``Call`` instruction has a special parameter handling logic.
+            This instruction separately keeps program, i.e. parametrized schedule,
+            and bound parameters until execution. The parameter assignment operation doesn't
+            immediately override its operand data.
+        """
+        new_table = copy(node.arguments)
+
+        for parameter, value in new_table.items():
+            if isinstance(value, ParameterExpression):
+                new_table[parameter] = self._assign_parameter_expression(value)
+        node._arguments = new_table
```

Comment:
this is also sketchy

Reply:
This is on purpose, because parameters are assigned just before execution to reuse a subroutine for different parameters. Perhaps you suggest to add arguments setter?

Reply:
I like the idea of an arguments setter. Makes it easier to handle Schedule changes in the future.